### PR TITLE
Switch to PostgreSQL

### PR DESCRIPTION
**Pull request by sasha370**
We need to switch to PostgreSQL and allow migrations.

Please be patient, as there are many small corrections and optimizations of database queries in the bottom PR.

### How to test
- copy default ENVs from `.env.example` to `.env`
- run `direnv allow`
- run `poetry install`
- run `docker composer up`
- run migrations `poetry run alembic upgrade head` 
- make tests according to instructions https://gist.github.com/sasha370/1c218f78bd0d2b2e41289b4814e0c778

### CODE REVIEW

File: models/qa_interaction.py [hunk 1]
```diff
@@ -17,6 +17,6 @@ class QAInteraction(Base):
     slack_user_id = Column(String)
     question_timestamp = Column(DateTime)
     answer_timestamp = Column(DateTime)
-    comments = Column(Text, default=json.dumps([]))
```
Comment (sasha370) [inline models/qa_interaction.py:20]: It is still Text field in Postgres

File: database/bookmarked_conversation_manager.py [hunk 2]
```diff
@@ -1,32 +1,16 @@
-from sqlalchemy.exc import SQLAlchemyError
 from models.bookmarked_conversation import BookmarkedConversation
 from datetime import datetime, timezone
-from database.database import Database
 
 
-class BookmarkedConversationManager:
```
Comment (sasha370) [inline database/bookmarked_conversation_manager.py:7]: I've removed class and left only module methods

File: vector/pages/retriever.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ def retrieve_relevant_ids(question: str, count: int) -> List[int]:
 
     # Extract and return the document IDs from the results
     if 'ids' in similar_items:
-        page_ids = [int(id) for sublist in similar_items['ids'] for id in sublist]
```
Comment (sasha370) [inline vector/pages/retriever.py:39]: SQL query failed when it is INT

File: slack/event_consumer.py [hunk 4]
```diff
@@ -47,61 +47,70 @@ def get_user_name_from_id(slack_web_client, user_id):
...
+                    ScoreManager(session).add_or_update_score(slack_user_id=question_event.user, category='seeker')
                     print(f"Score updated for user {question_event.user}")
                 except Exception as e:
                     print(f"Error updating score for user {question_event.user}: {e}")
                 self.web_client.chat_postMessage(channel=channel_id, text=response_text, thread_ts=message_ts)
                 print(f"\nResponse posted to Slack thread: {message_ts}\n")
             except Exception as e:
-                print(f"Error registering message as processed, adding to db and responding to the question on slack: {e}")
+                print(
+                    f"Error registering message as processed, adding to db and responding to the question on slack: {e}")
```
Comment (id-ilych) [inline slack/event_consumer.py:97]:
```suggestion
                print(f"Error registering message as processed, adding to db and responding to the question on slack: {e}")
```

File: vector/interactions/embeddings/generate_one.py [hunk 5]
```diff
@@ -50,12 +51,13 @@ def format_interaction(interaction):
...
-        QAInteractionManager().add_embed_to_interaction(interaction_id, embed)
-        logging.info(f"Interaction with ID {interaction_id} vectorized and stored in the database.")
-    else:
-        logging.error(f"No interaction found for ID {interaction_id}")
-    return None
+    with get_db_session() as session:
+        interaction = QAInteractionManager(session).get_interaction_by_interaction_id(interaction_id)
+        if interaction:
+            formatted_interaction = format_interaction(interaction)
+            embed = embed_text(formatted_interaction, model=embedding_model_id)
```
Comment (id-ilych) [inline vector/interactions/embeddings/generate_one.py:58]: That is time-consuming operation (network call), are you sure the session should be open for all that time?

File: pyproject.toml [hunk 6]
```diff
@@ -26,6 +26,9 @@ matplotlib = "^3.8.3"
 seaborn = "^0.13.2"
 plotly = "^5.20.0"
 umap-learn = "^0.5.5"
+jupyter = "^1.0.0"
+psycopg2 = "^2.9.9"
```
[the remark this answers is not in the document]
Comment (sasha370) [reply pyproject.toml:30]: This is driver for psql in SQLAlchemy

File: slack/event_consumer.py [hunk 7]
```diff
@@ -47,61 +47,70 @@ def get_user_name_from_id(slack_web_client, user_id):
...
             try:
-                self.add_question_and_response_to_database(question_event, response_text, assistant_thread_id=assistant_thread_id)
+                self.add_question_and_response_to_database(session,
+                                                           question_event,
+                                                           response_text,
+                                                           assistant_thread_id=assistant_thread_id)
                 try:
-                    self.score_manager.add_or_update_score(slack_user_id=question_event.user, category='seeker',
-                                                           points=1)
+                    ScoreManager(session).add_or_update_score(slack_user_id=question_event.user, category='seeker')
```
[the remark this answers is not in the document]
Comment (sasha370) [reply slack/event_consumer.py:89]: There is a default value eq 1 
https://github.com/toptal/top-assist/blob/511d450518e527729fe1c02f0e85303a36c083f3/database/score_manager.py#L12

File: database/quiz_question_manager.py [hunk 8]
```diff
@@ -1,75 +1,48 @@
...
-                question = session.query(QuizQuestion).filter_by(id=question_id).first()
-                if question:
-                    question.summary = summary
-                    session.commit()
-        except SQLAlchemyError as e:
-            print(f"Error updating quiz question with summary: {e}")
+        with get_db_session() as session:
+            new_question = QuizQuestion(question_text=question_text, posted_on_slack=datetime.now(timezone.utc))
+            session.add(new_question)
+            session.flush()
```
Comment (sasha370) [inline database/quiz_question_manager.py:15]: session.flush() in SQLAlchemy sends all accumulated changes (inserts, updates, deletes) to the database but keeps the transaction open. 
We need to get `new_question.id` before continue

File: slack/message_manager.py [hunk 9]
```diff
@@ -51,7 +49,7 @@ def post_questions_to_slack(channel_id, quiz_question_dtos, user_ids):
 
             # Award points to each revealer
             for user_id in user_ids:
-                score_manager.add_or_update_score(user_id, category='revealer', points=1)
+                ScoreManager().add_or_update_score(user_id, category='revealer')
```
Comment (sasha370) [inline slack/message_manager.py:52]: We have  `points=1` as a default for `add_or_update_score`

File: vector/pages/retriever.py [hunk 10]
```diff
@@ -36,7 +36,7 @@ def retrieve_relevant_ids(question: str, count: int) -> List[int]:
...
+        page_ids = [id for sublist in similar_items['ids'] for id in sublist]
```
Comment (sasha370) [inline vector/pages/retriever.py:39]: We use Strings for `page_ids`.
It is better to switch to BigInteger, but we need to fix Chroma inputs/outputs as well

**Review by id-ilych — APPROVED** [review]

**Pull request merged**